certora/specs/EnumerableMap.spec

```
@@ -0,0 +1,312 @@
+│ Invariant: the value mapping is empty for keys that are not in the EnumerableMap.                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant noValueIfNotContained(bytes32 key)
+    !contains(key) => tryGet_value(key) == 0
+    {
+        preserved set(bytes32 otherKey, bytes32 someValue) {
+            require sanity();
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: All indexed keys are contained                                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant indexedContained(uint256 index)
+    contains(at_key(index))
+    {
+        preserved {
+            requireInvariant consistencyIndex(index);
+            requireInvariant consistencyIndex(to_uint256(length() - 1));
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: A value can only be stored at a single location                                                          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant atUniqueness(uint256 index1, uint256 index2)
+    index1 == index2 <=> at_key(index1) == at_key(index2)
+    {
+        preserved remove(bytes32 key) {
+            requireInvariant atUniqueness(index1, to_uint256(length() - 1));
+            requireInvariant atUniqueness(index2, to_uint256(length() - 1));
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: index <> value relationship is consistent                                                                │
+│                                                                                                                     │
+│ Note that the two consistencyXxx invariants, put together, prove that at_ and _indexOf are inverse of one another.  │
+│ This proves that we have a bijection between indices (the enumerability part) and keys (the entries that are set    │
+│ and removed from the EnumerableMap).                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant consistencyIndex(uint256 index)
+    _indexOf(at_key(index)) == index + 1
+    {
+        preserved remove(bytes32 key) {
+            requireInvariant consistencyIndex(to_uint256(length() - 1));
+        }
+    }
+
+invariant consistencyKey(bytes32 key)
+    at_key(to_uint256(_indexOf(key) - 1)) == key
```

Here I would be more concerned about `_indexOf(key)` being 0. Shouldn't we be writing the following?

```suggestion
    _indexOf(key) > 0 => at_key(to_uint256(_indexOf(key) - 1)) == key
```

Otherwise it seems writing `requireInvariant consistencyKey(key)` may in some cases be requiring a condition that shouldn't be assumed.

---

If `_indexOf(key)` is 0, `to_uint256(_indexOf(key) - 1)` should revert, and nothing is checked.

Indeed `requireInvariant consistencyKey(key)` requires that the key exist. This is clear to me, but might not be to everyone.

---

The same applies to `invariant consistencyIndex(uint256 index)`. If the index is not in scope, `at_key(index)` reverts

---

> `to_uint256(_indexOf(key) - 1)` should revert

What is the part that should revert exactly? I expect `0 - 1` to evaluate to `-1`, and `to_uint256(-1)` evaluates to `max_uint256`.

---

I guess `at_key(max_uint256)` is the part that reverts. Will change the way to propose it